zerver/views/upload.py

```diff
@@ -224,9 +241,52 @@ def serve_image_error(status: int, image_path: str) -> HttpResponseBase:
         patch_vary_headers(response, ("Accept",))
         return response
     if url_only:
+        # TODO: Does this need to handle thumbnails?
         url = generate_unauthed_file_access_url(path_id)
         return json_success(request, data=dict(url=url))
 
+    if thumbnail_format is not None:
+        # Check if this is something that we thumbnail at all
+        try:
+            image_attachment = ImageAttachment.objects.get(path_id=path_id)
+        except ImageAttachment.DoesNotExist:
+            return serve_image_error(404, "images/errors/image-not-exist.png")
+
+        # Validate that this is a potential thumbnail format
+        requested_format = BaseThumbnailFormat.from_string(thumbnail_format)
+        if requested_format is None:
+            return serve_image_error(404, "images/errors/image-not-exist.png")
+
+        rendered_formats = [StoredThumbnailFormat(**f) for f in image_attachment.thumbnail_metadata]
+
+        # We never generate animated versions if the input was still,
+        # so filter those out
+        if image_attachment.frames == 1:
+            potential_output_formats = [
+                thumbnail_format
+                for thumbnail_format in THUMBNAIL_OUTPUT_FORMATS
+                if not thumbnail_format.animated
+            ]
+        else:
+            potential_output_formats = THUMBNAIL_OUTPUT_FORMATS
+        if requested_format not in potential_output_formats:
+            if requested_format in rendered_formats:
+                # Not a _current_ format, but we did render it at the time, so fine to serve
+                pass
+            else:
+                return serve_image_error(404, "images/errors/image-not-exist.png")
+        elif requested_format not in rendered_formats:
+            # They requested a valid format, but one we've not
+            # rendered yet.  Take a lock on the row, then render every
+            # missing format, synchronously.
```

Comment:
Maybe worth spelling out here that this is how we ensure we only render it once if several clients request the same thumbnail at once.